### DISCUSSION TITLE
Difference in behavior between ref.invalidate and ref.refresh

### DISCUSSION THREAD

**Discussion by motucraft** [rrousselGit/riverpod#2343] [Q&A]
Hi there.

Until recently, my understanding was that if I wanted a return value, I would use refresh, otherwise I would use invalidate.
That was all I was aware of.

But I noticed a difference. Is it related to copyWithPrevious, which maintains the previous state?
In a nutshell, they are as follows.

In the case of refresh, isLoading is not true. What is the difference intended?

- ref.invalidate
when method error callback is executed twice.
```bash
flutter: isLoading=false
flutter: isLoading=true
```
- ref.refresh
This is one time.
```bash
flutter: isLoading=false
```
- sample
```dart
import 'package:flutter/material.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'main.g.dart';

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Sample(),
    );
  }
}

class Sample extends ConsumerWidget {
  const Sample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final sampleAsyncValue = ref.watch(sampleNotifierProvider);

    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            sampleAsyncValue.when(
              data: (data) {
                return const Text('data');
              },
              error: (error, stack) {
                print('isLoading=${sampleAsyncValue.isLoading}');
                ref.invalidate(sampleNotifierProvider);
                // ref.refresh(sampleNotifierProvider);

                return const Text('error');
              },
              loading: () => const CircularProgressIndicator(),
            ),
            ElevatedButton(
              onPressed: () => ref.read(sampleNotifierProvider.notifier).throwException(),
              child: const Text('Error Button'),
            ),
          ],
        ),
      ),
    );
  }
}

@riverpod
class SampleNotifier extends _$SampleNotifier {
  @override
  FutureOr<String?> build() async {
    return null;
  }

  Future<void> throwException() async {
    state = const AsyncLoading();
    state = await AsyncValue.guard(() async => throw Exception('sample'));
  }
}

```

**Comment by motucraft** [comment]
It's hard to understand for me, so I'm trying to draw a diagram, but my understanding is something like this.
Is this understanding correct?


![image](https://user-images.githubusercontent.com/35750184/226084235-297aae08-fe8a-407a-9458-738fe4068394.png)

**Comment by rrousselGit** [comment]
There is something fundamentally wrong with your example. You're not allowed to call `ref.invalide` while the widget tree is building.

Your code shouldn't even render anything and you should get an assert error. It's probably an edge-case I forgot to handle. In any case, you shouldn't do that.

**Reply by motucraft** [reply]
@rrousselGit 
I see, that's a bad example. I did not get any assert error, so I thought there was no problem.
I'll be careful, thanks.

What if modify the sample like this?
There is a difference between ref.invalidate and ref.refresh when "Error Clear Button" is pressed after "Error Button".
In the case of ref.invalidate, the error callback is called and "flutter: isLoading=true" is output.
In contrast, in the case of "ref.refresh", the error callback is not invoked.

I don't know the difference.

```dart
import 'package:flutter/material.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'main.g.dart';

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Sample(),
    );
  }
}

class Sample extends ConsumerWidget {
  const Sample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final sampleAsyncValue = ref.watch(sampleNotifierProvider);

    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            sampleAsyncValue.when(
              data: (data) {
                return const Text('data');
              },
              error: (error, stack) {
                print('isLoading=${sampleAsyncValue.isLoading}');

                return const Text('error');
              },
              loading: () => const CircularProgressIndicator(),
            ),
            ElevatedButton(
              onPressed: () => ref.read(sampleNotifierProvider.notifier).throwException(),
              child: const Text('Error Button'),
            ),
            const SizedBox(height: 24),
            ElevatedButton(
              onPressed: () {
                ref.invalidate(sampleNotifierProvider);
                // ref.refresh(sampleNotifierProvider);
              },
              child: const Text('Error Clear Button'),
            ),
          ],
        ),
      ),
    );
  }
}

@riverpod
class SampleNotifier extends _$SampleNotifier {
  @override
  FutureOr<String?> build() async {
    return null;
  }

  Future<void> throwException() async {
    state = const AsyncLoading();
    state = await AsyncValue.guard(() async {
      await Future.delayed(const Duration(seconds: 3));
      throw Exception('sample');
    });
  }
}

```

**Reply by rrousselGit** [reply]
refresh is quite litteraly:

```dart
T refresh<T>(ProviderBase<T> p) {
  invalidate(p);
  return read(p);
}
```

There's no difference outside of this.

The only thing is that reading the provider causes an immediate refresh. Whereas not reading it delays the refresh to have it happen naturally on its own

**Reply by motucraft** [reply]
Thanks for sparing your precious time.

The API reference also states _"As opposed to refresh, the refresh is not immediate and is instead delayed to the next read or next frame"_. But I didn't understand what the behavior would change as a result.

I understood that the change illustrated  [here(#2343 (comment))](https://github.com/rrousselGit/riverpod/discussions/2343#discussioncomment-5351358) is produced by _"reading the provider causes an immediate refresh"_. 
Is my understanding still wrong?

**Reply by rrousselGit** [reply]
There's not really a difference in output.
If there is in your case then it could be because of some race conditions. But generally the output of both of these is the same

**Reply by motucraft** [reply]
Are you saying the no difference in terms of Eventual Consistency?
If so, it is understandable.

**Comment by ChandaLoehrsE2** [comment]
This post is useful for developers who are using Riverpod for Flutter application development.

**Comment by tonnyavery** [comment]
If `invalidate` takes time and next call to `read` hits FamilyProvider version that is not yet invalidated there in an error.
Any way to wait for `invalidate` to finish or are we supposed to use `refresh` instead?

**Reply by rrousselGit** [reply]
> next call to read hits FamilyProvider version that is not yet invalidated there in an error.

that shouldn't be. What's the error?

**Reply by tonnyavery** [reply]
Likely my example is too complex.

I have a `DataStoreProvider` that "calls into a database" (mocked for now) to store items of Todo list.
```
final todoDataStoreProvider = Provider((ref) => ToDoDataStore());
class ToDoDataStore {
  final List<ToDo> _todoDatabase = [];

  Future<List<ToDo>> get() async {
    return [..._todoDatabase];
  }

  Future<ToDo> getToDo(String todoId) async {
    return _todoDatabase.firstWhere((element) => element.id == todoId);
  }

  add(todo)
  remove(todo)
```

and a Future family provider to read single DataStore item

```
final singleToDoProvider =
    FutureProvider.family<ToDo, String>((ref, todoId) async {
  return await ref.read(todoDataStoreProvider).getToDo(todoId);
});
```

Just adding items to my TodoDataStore doesn't seem to effect `singleToDoProvider` so it can't find an item failing with `Bad state: No element`.

I have changed all usages from `read(singleToDoProvider)` to `watch(singleToDoProvider)`, but it doesn't seem to help.

To workaround `invalidate` taking time an not being `async` (so no way to wait for invalidation to finish), I added a call to `ref.refresh(singleToDoProvider(newTodoId));` after adding. 
This helped in some cases, but in others, where more providers are involved, I still get `Bad state: No element`.

It looks as if FamilyProviders that depend on `singleToDoProvider` also require invalidation any time I am changing some data in ToDoDataStore. Having many chains of providers depending on each other makes it difficult to find an invalidation path.
